### PR TITLE
Support open deployments and related realtime events

### PR DESCRIPTION
**Pull request by dennari**
If a raw deployment is considered 'open', it's served without authentication or authorization.

Adds a new endpoint at `/events/deployment/{projectId}-{deploymentId}` which returns `COMMENT_ADDED` and `COMMENT_DELETED` events related to the specified *open* deployment.

NOTE: change the base to master after #174 has been merged.

### CODE REVIEW

**Comment by vsaarinen** [comment]
If the user is not logged in, they don't have a teamId. If there's something that I'm missing, the path can't have `teamId`  in it.

File: src/realtime/realtime-hapi-plugin.ts [hunk 1]
```diff
@@ -146,45 +157,54 @@ export class RealtimeHapiPlugin {
...
-        nodeStream.push(null);
-      });
-
-    } catch (err) {
-      this.logger.error('Error handling a SSE request', err);
-    }
+  private requestHandlerFactory(streamFactory: StreamFactory) {
+    return async (request: Hapi.Request, reply: Hapi.IReply) => {
+      try {
+        // const observable = await this.getObservable(teamId, since ? parseInt(since, 10) : undefined);
```
Comment (vsaarinen) [inline src/realtime/realtime-hapi-plugin.ts]: Comment

File: src/realtime/realtime-hapi-plugin.ts [hunk 2]
```diff
@@ -304,7 +324,7 @@ export class RealtimeHapiPlugin {
     try {
       const deployment = event.payload.deployment;
       const apiResponse = await this.jsonApiPlugin.getEntity(
-          'deployment', api => api.toApiDeployment(deployment.projectId, deployment));
+        'deployment', api => api.toApiDeployment(deployment.projectId, deployment));
```
Comment (vsaarinen) [inline src/realtime/realtime-hapi-plugin.ts]: The last `)` should be on the following line.

File: src/realtime/realtime-hapi-plugin.ts [hunk 3]
```diff
@@ -66,36 +68,29 @@ import {
...
+    @inject(eventBusInjectSymbol) private readonly eventBus: PersistentEventBus,
+    @inject(logger.loggerInjectSymbol) private readonly logger: logger.Logger,
+  ) {
+    super({
+        name: 'realtime-plugin',
+        version: '1.0.0',
+    });
     this.persistedEvents = this.eventBus.getStream()
       .filter(isPersistedEvent)
       .map(event => <PersistedEvent<any>> event)
```
Comment (vsaarinen) [inline src/realtime/realtime-hapi-plugin.ts]: Since `isPersistedEvent` is a type guard, isn't this `map` unnecessary?

File: src/realtime/realtime-hapi-plugin.ts [hunk 4]
```diff
@@ -146,45 +157,54 @@ export class RealtimeHapiPlugin {
...
-        this.persistedEvents.filter(event => event.teamId === teamId),
-      );
-      if (since) {
-        const existing = await this.eventBus.getEvents(teamId, since);
-        if (existing.length > 0) {
-          existing.shift(); // getEvents is '>= since', but here we want '> since'
-        }
-        observable = Observable.concat(Observable.from(existing), observable);
+  private async getStream(request: Hapi.Request) {
+    const {teamId, projectId, deploymentId} = request.params;
```
Comment (vsaarinen) [inline src/realtime/realtime-hapi-plugin.ts]: Should have spaces inside curly brackets.

File: src/realtime/realtime-hapi-plugin.ts [hunk 5]
```diff
@@ -146,45 +157,54 @@ export class RealtimeHapiPlugin {
...
+      const _projectId = parseInt(projectId, 10);
+      const _deploymentId = parseInt(deploymentId, 10);
+      predicate = deploymentEventFilter(_teamId, _projectId, _deploymentId);
+    }
+    let observable = Observable.concat(
+      Observable.of(this.pingEvent()),
+      this.persistedEvents.filter(predicate),
+    );
+    if (since && !isNaN(since)) {
+      const existing = await this.eventBus.getEvents(_teamId, since);
```
Comment (vsaarinen) [inline src/realtime/realtime-hapi-plugin.ts]: Will this pass on missed events that are not related to this deployment if `projectId` and `deploymentId` exist?

File: src/realtime/realtime-hapi-plugin.ts (hunk 3 above)
Comment (dennari) [inline src/realtime/realtime-hapi-plugin.ts]: It's purpose is only to make the compiler agree that `this.persistedEvents` has the desired narrowed type.

File: src/realtime/realtime-hapi-plugin.ts (hunk 4 above)
Comment (dennari) [inline src/realtime/realtime-hapi-plugin.ts]: Are we missing some linter rule?

File: src/realtime/realtime-hapi-plugin.ts (hunk 5 above)
Comment (dennari) [inline src/realtime/realtime-hapi-plugin.ts]: A nice catch indeed 👍

**Comment by dennari** [comment]
Changed to not require the `teamId`. This PR now also includes the related changes for serving open raw deployments.

File: src/realtime/realtime-hapi-plugin.ts (hunk 3 above)
Comment (vsaarinen) [inline src/realtime/realtime-hapi-plugin.ts]: I tested it without the map and that seemed to be the case.

File: src/realtime/realtime-hapi-plugin.ts (hunk 4 above)
Comment (vsaarinen) [inline src/realtime/realtime-hapi-plugin.ts]: It's currently not a supported rule in tslint AFAIK.

Comment (vsaarinen) [inline src/realtime/realtime-hapi-plugin.ts]: https://github.com/palantir/tslint/issues/1044

File: src/realtime/realtime-hapi-plugin.ts [hunk 6]
```diff
@@ -112,202 +52,87 @@ export class RealtimeHapiPlugin {
...
+      handler: {
+        async: this.deploymentHandler,
+      },
+      config: {
+        bind: this,
+        auth: false,
+        cors: true,
+        validate: {
+          params: {
+            teamId: Joi.number().required(),
```
Comment (vsaarinen) [inline src/realtime/realtime-hapi-plugin.ts]: Should this be here?

File: src/realtime/realtime-hapi-plugin.ts [hunk 7]
```diff
@@ -112,202 +52,87 @@ export class RealtimeHapiPlugin {
...
-        commit: toApiCommitId(deployment.projectId, deployment.commitHash),
-        deployment: apiResponse.data,
-      };
-      return this.toSSE(event, ssePayload);
-    } catch (error) {
-      const msg = 'Could not convert DeploymentEvent to streaming event';
-      this.logger.error(msg, { error, event });
-      throw Error(msg);
+      const { projectId, deploymentId } = request.params;
+      const teamId = await request.getOpenTeamId();
```
Comment (vsaarinen) [inline src/realtime/realtime-hapi-plugin.ts]: Doesn't this now assume that whenever we fetch events from this endpoint that they belong to the open team? This is not the case if I'm a logged in user from another team and I use a Deployment View as a landing page, at which point this endpoint will be used.

File: src/server/hapi.ts [hunk 8]
```diff
@@ -20,6 +20,8 @@ declare module 'hapi' {
   interface RequestDecorators {
     userHasAccessToProject: (projectId: number) => Promise<boolean>;
     userHasAccessToTeam: (teamId: number) => Promise<boolean>;
+    isOpenDeployment: (projectId: number, deploymentId: number) => Promise<boolean>;
+    getOpenTeamId: () => Promise<number | undefined>;
```
Comment (vsaarinen) [inline src/server/hapi.ts]: What if we have multiple teams that want open deployments?

File: src/realtime/realtime-hapi-plugin.ts (hunk 7 above)
Comment (vsaarinen) [inline src/realtime/realtime-hapi-plugin.ts]: A higher-level comment about this handler in the main review comment.

File: src/server/hapi.ts (hunk 8 above)
Comment (vsaarinen) [inline src/server/hapi.ts]: I'd maybe get rid of this function altogether. More info in the main review comment.

File: src/deployment/deployment-hapi-plugin.ts [hunk 9]
```diff
@@ -41,117 +39,122 @@ class DeploymentHapiPlugin extends HapiPlugin {
...
-            },
-          },
-        },
-      });
+      server.route(this.gitlabYmlRoute());
+      server.route(this.rawDeploymentRoutes(false));
     } else {
+      server.route(this.traceRoute());
+      server.route(this.rawDeploymentRoutes(false));
+      server.route(this.rawDeploymentRoutes(true));
```
Comment (dennari) [inline src/deployment/deployment-hapi-plugin.ts]: Yes, the one with `false` is for the open deployments

File: src/realtime/realtime-module.ts [hunk 10]
```diff
@@ -0,0 +1,245 @@
...
+  ) {
+
+    // creates SSEEvents and posts them
+    this.eventBusSubscription = this.getEnrichedStream(this.eventBus.getStream())
+      .subscribe(this.eventBus.post.bind(this.eventBus));
+
+    // listens to SSEEvents and shares them
+    this.sseStream = this.eventBus.getStream()
+      .filter(isPersistedEvent)
+      .map(event => event as PersistedEvent<any>)
```
Comment (dennari) [inline src/realtime/realtime-module.ts]: Right. I'd like to say the typings for `filter` have improved, making the map unnecessary.

**Comment by dennari** [comment]
Good points. The initial implementation left much to be desired. I've now rehauled the way the 'open' check works and feel this is much closer to what it should've been.

File: src/integration-test/system-integration-tests.ts [hunk 11]
```diff
@@ -39,6 +39,7 @@ interface SSE {
 
 const flowToken = process.env.FLOWDOCK_FLOW_TOKEN;
 const projectFolder = process.env.SYSTEM_TEST_PROJECT ? process.env.SYSTEM_TEST_PROJECT : 'blank';
+const openProjectFolder = process.env.SYSTEM_TEST_PROJECT_OPEN ? process.env.SYSTEM_TEST_PROJECT_OPEN : 'blank-open';
```
Comment (dennari) [inline src/integration-test/system-integration-tests.ts]: Done

File: src/authentication/authentication-hapi-plugin.ts [hunk 12]
```diff
@@ -264,16 +296,16 @@ class AuthenticationHapiPlugin extends HapiPlugin {
...
 
-  private validateFuncFactory(authorizer: (userName: string, request: Hapi.Request) => Promise<boolean>) {
+  private validateFuncFactory(authorizer: Authorizer) {
     return async (
       payload: AccessToken,
       request: Hapi.Request,
       callback: (err: any, valid: boolean, credentials?: any) => void,
     ) => {
-      let isAuthorized = false;
+      let isAuthorized: boolean | undefined = false;
```
Comment (vsaarinen) [inline src/authentication/authentication-hapi-plugin.ts]: Since this can now have three values, I'd prefer e.g. an enum to make it explicit what each state means.

File: src/authentication/authentication-hapi-plugin.ts [hunk 13]
```diff
@@ -305,28 +315,27 @@ class AuthenticationHapiPlugin extends HapiPlugin {
...
+      if (authorizationStatus === AuthorizationStatus.UNAUTHORIZED) {
         this.logger.debug('User %s not is not authorized', payload.username || payload.sub);
       }
-      // isAuthorized === undefined means the user was authenticated
-      // but the authorization hasn't been checked
       return callback(
         undefined,
-        isAuthorized === true || isAuthorized === undefined,
+        authorizationStatus === AuthorizationStatus.AUTHORIZED ||
+        authorizationStatus === AuthorizationStatus.NOT_CHECKED,
```
Comment (vsaarinen) [inline src/authentication/authentication-hapi-plugin.ts]: Should be indented a bit more

File: src/deployment/deployment-hapi-plugin.ts [hunk 14]
```diff
@@ -41,117 +38,96 @@ class DeploymentHapiPlugin extends HapiPlugin {
...
+  protected async onRequest(request: Hapi.Request, reply: Hapi.IReply) {
     // hostname doesn't include the port, host does
     const { hostname } = request.info;
-    if (isRawDeploymentHostname(hostname)) {
+    const key = getDeploymentKeyFromHost(hostname);
+    if (key !== null) {
       // Store the original hostname
       request.app.hostname = hostname;
       // Overwrite the hostname to use virtual host routing
+      // Each vhost (including the default one) has a separate routing table
```
Comment (vsaarinen) [inline src/deployment/deployment-hapi-plugin.ts]: Is this still relevant?

File: src/authentication/authentication-hapi-plugin.ts [hunk 15]
```diff
@@ -256,38 +296,48 @@ class AuthenticationHapiPlugin extends HapiPlugin {
...
       }
-      if (!isAuthorized) {
+      if (authorizationStatus === AuthorizationStatus.UNAUTHORIZED) {
         this.logger.debug('User %s not is not authorized', payload.username || payload.sub);
       }
-      return callback(undefined, isAuthorized, payload);
+      return callback(
+        undefined,
+        authorizationStatus === AuthorizationStatus.AUTHORIZED ||
+        authorizationStatus === AuthorizationStatus.NOT_CHECKED,
```
Comment (vsaarinen) [inline src/authentication/authentication-hapi-plugin.ts]: Should be indented a bit more.

File: src/deployment/deployment-hapi-plugin.ts (hunk 14 above)
Comment (dennari) [inline src/deployment/deployment-hapi-plugin.ts]: Yes, we still use two separate ones

**Comment by vsaarinen** [comment]
:tada:
